AlphaWallet/Common/Services/ContentLoader.swift

```diff
@@ -0,0 +1,233 @@
+                            } catch {
+                                return .fail(ContentLoaderError.invalidData)
+                            }
+                        }
+                        .handleEvents(receiveCompletion: { _ in self?.publishers[urlRequest] = nil })
+                        .share()
+                        .receive(on: RunLoop.main)
+                        .eraseToAnyPublisher()
+
+                    self?.publishers[urlRequest] = publisher
+
+                    return publisher
+                }
+            }.eraseToAnyPublisher()
+    }
+}
+
+extension ContentLoader {
+    typealias LoadContentPublisher = AnyPublisher<Loadable<Content, NoError>, ContentLoaderError>
+
+    struct NoError: Error {}
+
+    enum ContentLoaderError: Error {
+        case `internal`(Error)
+        case invalidData
+        case invalidUrl
+    }
+
+    enum Content {
+        case image(UIImage)
+        case svg(String)
+        case video(Video)
+    }
+
+    struct Video {
+        let url: URL
+        let preview: UIImage?
+    }
+
+    private struct VideoUrl: Codable {
+        let url: URL
+    }
+
+    struct ImageDecoder {
+        typealias Response = (data: Data, image: Content)
+
+        private func isValidResponse(statusCode: Int) -> Bool {
+            return (200...299).contains(statusCode)
+        }
+
+        func decode(response: HTTPURLResponse, data: Data) throws -> Response {
+            guard let url = response.url else { throw DecoderError.responseUrlNotFound }
+            guard isValidResponse(statusCode: response.statusCode) else { throw DecoderError.invalidStatusCode(response.statusCode) }
+            guard !data.isEmpty else { throw DecoderError.emptyDataResponse }
+
+            return try decode(data: data, url: url)
+        }
+
+        enum DecoderError: Error {
+            case decodeFailure(data: Data)
+            case emptyDataResponse
+            case responseUrlNotFound
+            case invalidStatusCode(Int)
+        }
+
+        func decode(data: Data) throws -> Content {
+            if let image = UIImage(data: data) {
+                return .image(image)
+            } else if let data = try? JSONDecoder().decode(VideoUrl.self, from: data) {
+                return .video(.init(url: data.url, preview: nil))
+            } else if let svg = String(data: data, encoding: .utf8) {
+                return .svg(svg)
+            } else {
+                throw DecoderError.decodeFailure(data: data)
+            }
+        }
+
+        private func decode(data: Data, url: URL) throws -> Response {
+            if let image = UIImage(data: data) {
+                return (data: data, .image(image))
+            } else if let svg = String(data: data, encoding: .utf8) {
+                return (data: data, .svg(svg))
+            } else {
+                let data = try JSONEncoder().encode(VideoUrl(url: url))
```

Comment:
So we don't cache the video's data like we do for images and SVGs right? Do we not do it because of size? But does it mean if user taps a token and load the video, dismiss it and taps the token again, the video might be downloaded again?

If we don't cache it to disk, should we cache it to memory with `ImageCache`? Does `ImageCache` purge the memory cache properly?

Reply:
for now yes, its gonna to be downloaded again, only if caching not applied inside on AVAsset